File: python/features/steps/jira_sync_steps.py
"""Behave steps for Jira pull synchronization."""

from __future__ import annotations

import json
import socket
import threading
from http.server import BaseHTTPRequestHandler, HTTPServer
from pathlib import Path
from typing import Any, Dict, List
from urllib.parse import urlparse, parse_qs

from behave import given, then, when

from features.steps.shared import load_project_directory, run_cli


def _find_free_port() -> int:
    """Find a free TCP port on localhost.

    :return: Available port number.
    :rtype: int
    """
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
        sock.bind(("127.0.0.1", 0))
        return sock.getsockname()[1]


def _build_jira_issue(
    key: str,
    summary: str,
    issue_type: str,
    status: str,
    priority: str,
    parent_key: str = "",
) -> Dict[str, Any]:
    """Build a minimal Jira issue JSON payload.

    :param key: Jira issue key.
    :type key: str
    :param summary: Issue summary.
    :type summary: str
    :param issue_type: Jira issue type name.
    :type issue_type: str
    :param status: Jira status name.
    :type status: str
    :param priority: Jira priority name.
    :type priority: str
    :param parent_key: Optional parent Jira key.
    :type parent_key: str
    :return: Jira issue dict.
    :rtype: Dict[str, Any]
    """
    issue: Dict[str, Any] = {
        "key": key,
        "fields": {
            "summary": summary,
            "description": None,
            "issuetype": {"name": issue_type},
            "status": {"name": status},
            "priority": {"name": priority},
            "assignee": None,
            "reporter": {"displayName": "test-reporter"},
            "labels": [],
            "comment": {"comments": []},
            "created": "2026-01-01T00:00:00.000+0000",
            "updated": "2026-01-01T00:00:00.000+0000",
            "resolutiondate": None,
        },
    }
    if parent_key:
        issue["fields"]["parent"] = {"key": parent_key}
    return issue


def _start_fake_jira_server(issues: List[Dict[str, Any]], port: int) -> HTTPServer:
    """Start a fake Jira HTTP server on the given port.

    :param issues: List of Jira issue dicts to serve.
    :type issues: List[Dict[str, Any]]
    :param port: Port to listen on.
    :type port: int
    :return: Running HTTPServer instance.
    :rtype: HTTPServer
    """
    issues_snapshot = list(issues)

    class FakeJiraHandler(BaseHTTPRequestHandler):
        def do_GET(self) -> None:
            parsed = urlparse(self.path)
            if parsed.path == "/rest/api/3/search/jql":
                params = parse_qs(parsed.query)
                start_at = int(params.get("startAt", ["0"])[0])
                max_results = int(params.get("maxResults", ["100"])[0])
                page = issues_snapshot[start_at : start_at + max_results]
                response = json.dumps(
                    {
                        "issues": page,
                        "total": len(issues_snapshot),
                        "startAt": start_at,
                        "maxResults": max_results,
                    }
                ).encode()
                self.send_response(200)
                self.send_header("Content-Type", "application/json")
                self.end_headers()
                self.wfile.write(response)
            else:
                self.send_response(404)
                self.end_headers()

        def log_message(self, format: str, *args: object) -> None:
            pass

    server = HTTPServer(("127.0.0.1", port), FakeJiraHandler)
    thread = threading.Thread(target=server.serve_forever, daemon=True)
    thread.start()
    return server


def _parse_issues_table(context: object) -> List[Dict[str, Any]]:
    """Parse the Gherkin table of issues from context.

    :param context: Behave context object.
    :type context: object
    :return: List of Jira issue dicts.
    :rtype: List[Dict[str, Any]]
    """
    issues = []
    for row in context.table:
        issues.append(
            _build_jira_issue(
                key=row["key"],
                summary=row["summary"],
                issue_type=row["type"],
                status=row["status"],
                priority=row["priority"],
                parent_key=row.get("parent", ""),
            )
        )
    return issues


@given("a fake Jira server is running with issues:")
def given_fake_jira_server(context: object) -> None:
    """Start a fake Jira HTTP server with the issues from the table.

    :param context: Behave context object.
    :type context: object
    """
    issues = _parse_issues_table(context)
    port = _find_free_port()
    server = _start_fake_jira_server(issues, port)
    context.fake_jira_server = server
    context.fake_jira_port = port
    context.fake_jira_issues = issues


@given("the Kanbus configuration includes Jira settings pointing at the fake server")
def given_jira_config(context: object) -> None:
    """Write a jira: block into the project .kanbus.yml.

    :param context: Behave context object.
    :type context: object
    """
    port = context.fake_jira_port
    config_path = Path(context.working_directory) / ".kanbus.yml"
    existing = config_path.read_text(encoding="utf-8")
    jira_config = (
        "\njira:\n"
        f"  url: http://127.0.0.1:{port}\n"
        "  project_key: AQ\n"
        "  sync_direction: pull\n"
        "  type_mappings:\n"
        "    Task: task\n"
        "    Bug: bug\n"
        "    Workstream: epic\n"
    )
    config_path.write_text(existing + jira_config, encoding="utf-8")
    context.environment_overrides["JIRA_API_TOKEN"] = "test-token"
    context.environment_overrides["JIRA_USER_EMAIL"] = "test@example.com"


@given('the environment variable "{name}" is unset')
def given_env_var_unset(context: object, name: str) -> None:
    """Remove an environment variable from the test overrides.

    :param context: Behave context object.
    :type context: object
    :param name: Environment variable name.
    :type name: str
    """
    context.environment_overrides.pop(name, None)
    import os

    context._unset_env_vars = getattr(context, "_unset_env_vars", [])
    context._unset_env_vars.append((name, os.environ.get(name)))
    os.environ.pop(name, None)


@when('I run "kanbus jira pull"')
@given('I run "kanbus jira pull"')
def when_run_jira_pull(context: object) -> None:
    """Run the jira pull command.

    :param context: Behave context object.
    :type context: object
    """
    run_cli(context, "kanbus jira pull")


@when('I run "kanbus jira pull --dry-run"')
def when_run_jira_pull_dry_run(context: object) -> None:
    """Run the jira pull command in dry-run mode.

    :param context: Behave context object.
    :type context: object
    """
    run_cli(context, "kanbus jira pull --dry-run")


@then("{count:d} issue files should exist in the issues directory")
def then_issue_file_count(context: object, count: int) -> None:
    """Assert the number of issue JSON files in the issues directory.

    :param context: Behave context object.
    :type context: object
    :param count: Expected number of issue files.
    :type count: int
    """
    project_dir = load_project_directory(context)
    issues_dir = project_dir / "issues"
    actual = len(list(issues_dir.glob("*.json"))) if issues_dir.exists() else 0
    assert actual == count, f"Expected {count} issue files, found {actual}"


@then('an issue file with jira_key "{jira_key}" should exist with title "{title}"')
def then_issue_file_exists_with_title(
    context: object, jira_key: str, title: str
) -> None:
    """Assert an issue file with the given jira_key and title exists.

    :param context: Behave context object.
    :type context: object
    :param jira_key: Jira issue key stored in custom field.
    :type jira_key: str
    :param title: Expected issue title.
    :type title: str
    """
    issue = _find_issue_by_jira_key(context, jira_key)
    assert issue is not None, f"No issue found with jira_key {jira_key!r}"
    assert issue["title"] == title, f"Expected title {title!r}, got {issue['title']!r}"


@then('an issue file with jira_key "{jira_key}" should have type "{issue_type}"')
def then_issue_has_type(context: object, jira_key: str, issue_type: str) -> None:
    """Assert an issue file has the given type.

    :param context: Behave context object.
    :type context: object
    :param jira_key: Jira issue key.
    :type jira_key: str
    :param issue_type: Expected Kanbus issue type.
    :type issue_type: str
    """
    issue = _find_issue_by_jira_key(context, jira_key)
    assert issue is not None, f"No issue found with jira_key {jira_key!r}"
    actual = issue.get("type") or issue.get("issue_type")
    assert actual == issue_type, f"Expected type {issue_type!r}, got {actual!r}"


@then(
    'the issue with jira_key "{child_key}" should have a parent matching'
    ' the issue with jira_key "{parent_key}"'
)
def then_issue_parent_matches(context: object, child_key: str, parent_key: str) -> None:
    """Assert the child issue's parent field matches the parent issue's Kanbus ID.

    :param context: Behave context object.
    :type context: object
    :param child_key: Jira key of the child issue.
    :type child_key: str
    :param parent_key: Jira key of the expected parent issue.
    :type parent_key: str
    """
    parent_issue = _find_issue_by_jira_key(context, parent_key)
    child_issue = _find_issue_by_jira_key(context, child_key)
    assert parent_issue is not None, f"No issue found with jira_key {parent_key!r}"
    assert child_issue is not None, f"No issue found with jira_key {child_key!r}"
    assert (
        child_issue.get("parent") == parent_issue["id"]
    ), f"Expected parent {parent_issue['id']!r}, got {child_issue.get('parent')!r}"


@then('the output should contain "{text}"')
def then_output_contains(context: object, text: str) -> None:
    """Assert that combined stdout+stderr contains the given text.

    :param context: Behave context object.
    :type context: object
    :param text: Expected text.
    :type text: str
    """
    import re as _re

    ansi_re = _re.compile(r"\x1b\[[0-9;]*m")
    combined = ansi_re.sub("", context.result.output or "")
    assert text in combined, f"Expected {text!r} in output, got:\n{combined}"


def _find_issue_by_jira_key(context: object, jira_key: str) -> Dict[str, Any] | None:
    """Find an issue JSON dict by its jira_key custom field.

    :param context: Behave context object.
    :type context: object
    :param jira_key: Jira issue key.
    :type jira_key: str
    :return: Issue dict or None.
    :rtype: Dict[str, Any] | None
    """
    project_dir = load_project_directory(context)
    issues_dir = project_dir / "issues"
    if not issues_dir.exists():
        return None
    for path in issues_dir.glob("*.json"):
        data = json.loads(path.read_text(encoding="utf-8"))
        custom = data.get("custom") or {}
        if custom.get("jira_key") == jira_key:
            return data
    return None

File: rust/features/steps/jira_sync_steps.rs
use std::collections::BTreeMap;
use std::net::TcpListener;
use std::sync::Arc;
use std::thread;
use std::time::Duration;

use axum::extract::Query;
use axum::response::IntoResponse;
use axum::routing::get;
use axum::Json;
use axum::Router;
use cucumber::{given, then};
use serde_json::{json, Value};
use serde_yaml::{Mapping, Value as YamlValue};
use tokio::sync::oneshot;

use crate::step_definitions::initialization_steps::KanbusWorld;

/// Allocate a free TCP port by binding to port 0 and reading the assigned port.
fn allocate_port() -> u16 {
    TcpListener::bind("127.0.0.1:0")
        .expect("bind ephemeral port")
        .local_addr()
        .expect("read local addr")
        .port()
}

/// Build a minimal Jira issue JSON value.
fn build_jira_issue(
    key: &str,
    summary: &str,
    issue_type: &str,
    status: &str,
    priority: &str,
    parent_key: Option<&str>,
) -> Value {
    let mut fields = json!({
        "summary": summary,
        "description": null,
        "issuetype": {"name": issue_type},
        "status": {"name": status},
        "priority": {"name": priority},
        "assignee": null,
        "reporter": {"displayName": "test-reporter"},
        "labels": [],
        "comment": {"comments": []},
        "created": "2026-01-01T00:00:00.000+0000",
        "updated": "2026-01-01T00:00:00.000+0000",
        "resolutiondate": null,
    });
    if let Some(parent) = parent_key {
        fields["parent"] = json!({"key": parent});
    }
    json!({"key": key, "fields": fields})
}

#[given("a fake Jira server is running with issues:")]
async fn given_fake_jira_server(world: &mut KanbusWorld, step: &cucumber::gherkin::Step) {
    let issues: Vec<Value> = step
        .table
        .as_ref()
        .map(|table| {
            let headers: Vec<&str> = table.rows[0].iter().map(|s| s.as_str()).collect();
            table.rows[1..]
                .iter()
                .map(|row_cells| {
                    let mut map: BTreeMap<String, String> = BTreeMap::new();
                    for (header, cell) in headers.iter().zip(row_cells.iter()) {
                        map.insert(header.to_string(), cell.clone());
                    }
                    let key = map.get("key").map(String::as_str).unwrap_or("");
                    let summary = map.get("summary").map(String::as_str).unwrap_or("");
                    let issue_type = map.get("type").map(String::as_str).unwrap_or("Task");
                    let status = map.get("status").map(String::as_str).unwrap_or("To Do");
                    let priority = map.get("priority").map(String::as_str).unwrap_or("Medium");
                    let parent = map
                        .get("parent")
                        .filter(|s| !s.is_empty())
                        .map(String::as_str);
                    build_jira_issue(key, summary, issue_type, status, priority, parent)
                })
                .collect()
        })
        .unwrap_or_default();

    let port = allocate_port();
    let issues = Arc::new(issues);
    let issues_clone = Arc::clone(&issues);

    let app = Router::new().route(
        "/rest/api/3/search/jql",
        get(move |Query(params): Query<BTreeMap<String, String>>| {
            let issues = Arc::clone(&issues_clone);
            async move {
                let start_at: usize = params
                    .get("startAt")
                    .and_then(|s| s.parse().ok())
                    .unwrap_or(0);
                let max_results: usize = params
                    .get("maxResults")
                    .and_then(|s| s.parse().ok())
                    .unwrap_or(100);
                let total = issues.len();
                let page: Vec<Value> = issues
                    .iter()
                    .skip(start_at)
                    .take(max_results)
                    .cloned()
                    .collect();
                Json(json!({
                    "issues": page,
                    "total": total,
                    "startAt": start_at,
                    "maxResults": max_results,
                }))
                .into_response()
            }
        }),
    );

    let (shutdown_tx, shutdown_rx) = oneshot::channel::<()>();

    thread::spawn(move || {
        let rt = tokio::runtime::Builder::new_current_thread()
            .enable_all()
            .build()
            .expect("tokio runtime");
        rt.block_on(async move {
            let listener = tokio::net::TcpListener::bind(format!("127.0.0.1:{port}"))
                .await
                .expect("bind fake jira");
            axum::serve(listener, app)
                .with_graceful_shutdown(async move {
                    let _ = shutdown_rx.await;
                })
                .await
                .expect("serve fake jira");
        });
    });

    // Give the server a moment to start accepting connections.
    thread::sleep(Duration::from_millis(50));

    world.fake_jira_port = Some(port);
    world.fake_jira_shutdown_tx = Some(shutdown_tx);
    world.fake_jira_issues = issues.as_ref().clone();
}

#[given("the Kanbus configuration includes Jira settings pointing at the fake server")]
fn given_jira_config(world: &mut KanbusWorld) {
    let port = world.fake_jira_port.expect("fake jira port not set");
    let config_path = world
        .configuration_path
        .clone()
        .unwrap_or_else(|| {
            world
                .working_directory
                .as_ref()
                .expect("working directory not set")
                .join(".kanbus.yml")
        });
    let existing = std::fs::read_to_string(&config_path).expect("read .kanbus.yml");
    let mut config_value: YamlValue =
        serde_yaml::from_str(&existing).expect("parse .kanbus.yml");
    let jira_value = {
        let mut jira = Mapping::new();
        jira.insert(
            YamlValue::String("url".to_string()),
            YamlValue::String(format!("http://127.0.0.1:{port}")),
        );
        jira.insert(
            YamlValue::String("project_key".to_string()),
            YamlValue::String("AQ".to_string()),
        );
        jira.insert(
            YamlValue::String("sync_direction".to_string()),
            YamlValue::String("pull".to_string()),
        );
        let mut type_mappings = Mapping::new();
        type_mappings.insert(
            YamlValue::String("Task".to_string()),
            YamlValue::String("task".to_string()),
        );
        type_mappings.insert(
            YamlValue::String("Bug".to_string()),
            YamlValue::String("bug".to_string()),
        );
        type_mappings.insert(
            YamlValue::String("Workstream".to_string()),
            YamlValue::String("epic".to_string()),
        );
        jira.insert(
            YamlValue::String("type_mappings".to_string()),
            YamlValue::Mapping(type_mappings),
        );
        YamlValue::Mapping(jira)
    };
    match &mut config_value {
        YamlValue::Mapping(mapping) => {
            mapping.insert(YamlValue::String("jira".to_string()), jira_value);
        }
        _ => {
            let mut mapping = Mapping::new();
            mapping.insert(YamlValue::String("jira".to_string()), jira_value);
            config_value = YamlValue::Mapping(mapping);
        }
    }
    let updated = serde_yaml::to_string(&config_value).expect("serialize .kanbus.yml");
    std::fs::write(&config_path, updated).expect("write .kanbus.yml");
    std::env::set_var("JIRA_API_TOKEN", "test-token");
    std::env::set_var("JIRA_USER_EMAIL", "test@example.com");
    world.jira_env_set = true;
    world.configuration_path = Some(config_path);
}

#[given(expr = "the environment variable {string} is unset")]
fn given_env_var_unset(world: &mut KanbusWorld, name: String) {
    let original = std::env::var(&name).ok();
    std::env::remove_var(&name);
    world.jira_unset_env_vars.push((name, original));
}

#[then(expr = "{int} issue files should exist in the issues directory")]
fn then_issue_file_count(world: &mut KanbusWorld, count: usize) {
    let issues_dir = world
        .working_directory
        .as_ref()
        .expect("working directory not set")
        .join("project")
        .join("issues");
    let actual = if issues_dir.exists() {
        std::fs::read_dir(&issues_dir)
            .expect("read issues dir")
            .filter_map(|e| e.ok())
            .filter(|e| {
                e.path()
                    .extension()
                    .and_then(|ext| ext.to_str())
                    .map(|ext| ext == "json")
                    .unwrap_or(false)
            })
            .count()
    } else {
        0
    };
    assert_eq!(
        actual, count,
        "Expected {count} issue files, found {actual}"
    );
}

#[then(expr = "an issue file with jira_key {string} should exist with title {string}")]
fn then_issue_exists_with_title(world: &mut KanbusWorld, jira_key: String, title: String) {
    let issue = find_issue_by_jira_key(world, &jira_key)
        .unwrap_or_else(|| panic!("No issue found with jira_key {jira_key:?}"));
    let actual_title = issue["title"].as_str().unwrap_or("");
    assert_eq!(
        actual_title, title,
        "Expected title {title:?}, got {actual_title:?}"
    );
}

#[then(expr = "an issue file with jira_key {string} should have type {string}")]
fn then_issue_has_type(world: &mut KanbusWorld, jira_key: String, issue_type: String) {
    let issue = find_issue_by_jira_key(world, &jira_key)
        .unwrap_or_else(|| panic!("No issue found with jira_key {jira_key:?}"));
    let actual = issue["type"].as_str().unwrap_or("");
    assert_eq!(
        actual, issue_type,
        "Expected type {issue_type:?}, got {actual:?}"
    );
}

#[then(
    expr = "the issue with jira_key {string} should have a parent matching the issue with jira_key {string}"
)]
fn then_issue_parent_matches(world: &mut KanbusWorld, child_key: String, parent_key: String) {
    let parent_issue = find_issue_by_jira_key(world, &parent_key)
        .unwrap_or_else(|| panic!("No issue found with jira_key {parent_key:?}"));
    let child_issue = find_issue_by_jira_key(world, &child_key)
        .unwrap_or_else(|| panic!("No issue found with jira_key {child_key:?}"));
    let expected_parent_id = parent_issue["id"].as_str().unwrap_or("");
    let actual_parent = child_issue["parent"].as_str().unwrap_or("");
    assert_eq!(
        actual_parent, expected_parent_id,
        "Expected parent {expected_parent_id:?}, got {actual_parent:?}"
    );
}

fn find_issue_by_jira_key(world: &KanbusWorld, jira_key: &str) -> Option<Value> {
    let issues_dir = world
        .working_directory
        .as_ref()?
        .join("project")
        .join("issues");
    if !issues_dir.exists() {
        return None;
    }
    for entry in std::fs::read_dir(&issues_dir).ok()?.flatten() {
        let path = entry.path();
        if path.extension().and_then(|e| e.to_str()) != Some("json") {
            continue;
        }
        if let Ok(content) = std::fs::read_to_string(&path) {
            if let Ok(data) = serde_json::from_str::<Value>(&content) {
                if data["custom"]["jira_key"].as_str() == Some(jira_key) {
                    return Some(data);
                }
            }
        }
    }
    None
}
